app/services/stage_update_service.rb

```diff
@@ -9,10 +9,50 @@ def update(options)
     old_state = stage.state
     stage.update_attributes(options)
     return if old_state == stage.state
-    if stage.state == Stage::NOTIFIED_STATE
+    case stage.state
+    when Stage::NOTIFIED_STATE
       EventService.new(stage.request).approver_group_notified(stage)
-    elsif stage.state == Stage::FINISHED_STATE
+      request_started if first_stage?
+    when Stage::FINISHED_STATE
       EventService.new(stage.request).approver_group_finished(stage)
+      request_finished(stage.decision, stage.reason) if last_stage?
+    when Stage::SKIPPED_STATE
+      last_stage_skipped if last_stage?
```

Comment:
If stage's status is marked as `denied`, all following stages should mark their states to `SKIPPED_STATE`, or just break and update request's state directly. I didn't see this logic.

How can stage have `Stage::SKIPPED_STATE`?

Reply:
See my comments above. BPM should set stage to skipped.

Reply:
In this case BPM will update stage via rest api client, we should keep those stage PATCH endpoints.

Reply:
Now I understand. BPM will create an action to update its stage's state.  Above comment can be ignored.